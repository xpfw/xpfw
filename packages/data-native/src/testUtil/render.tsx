import * as React from "react"
import rtr from "react-test-renderer"

export default (Component: any, name: string) => {
  const rendered = rtr.create(Component)
  expect(rendered.toJSON()).toMatchSnapshot(name)
  rendered.unmount()
}
