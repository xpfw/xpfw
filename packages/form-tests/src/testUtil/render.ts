import * as React from "react"
import * as render from "react-test-renderer"

const renderSnapshot = (component: any, msg: string) => {
  const tree = render.create(component)
  expect(tree.toJSON()).toMatchSnapshot(msg)
  tree.unmount()
}

export default renderSnapshot
