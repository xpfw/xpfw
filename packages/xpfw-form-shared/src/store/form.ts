import {
  globals, IField, IFieldError, IForm, IFormError, iterateFields,
  validateField, validateForm, ValidationRegistry
} from "@xpfw/validate"
import { get, isFunction, isNil, isObject, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"
import FormErrorStore from "./error"

@AutoSubscribeStore
export class FormStore extends StoreBase implements IPersistableStore {
  public name = "formData"
  private formData: any = {}

  public getPropKeys() { return ["formData"] }

  public setValue(valuePath: string, value: any, fullTrigger: boolean = false) {
    set(this.formData, valuePath, value)
    this.trigger(fullTrigger ? undefined : valuePath)
  }

  @autoSubscribe
  public getValue(@key valuePath: string): any {
    return get(this.formData, valuePath)
  }

  public getFormData(form: IForm, prefix: string = "", method: string = "create") {
    const fullData: any = {}
    const formData = this.formData
    iterateFields(form, (field) => {
      let defaultValue = get(field, "validate.defaultValue")
      if (method === globals.Method.Find) {
        defaultValue = undefined
      }
      let value = get(formData,
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, defaultValue)
      let converter: any = get(field, "validate.convert")
      if (!isFunction(converter) && isObject(converter)) {
        converter = converter[method]
      }
      if (isFunction(converter)) {
        value = converter(value)
      }
      if (method !== globals.Method.Find || method === globals.Method.Find && !isNil(value))  {
        set(fullData, field.mapTo, value)
      }
    })
    return fullData
  }

  public copyDocToFormData(doc: any, form: IForm, prefix: string = "") {
    iterateFields(form, (field) => {
        this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, get(doc, field.mapTo))
    })
    this.trigger()
  }

  public async validateField(field: IField, form: IForm, method: string = "create",
                             prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`
    try {
      await validateField(get(this.formData, valuePath), field, {
        method,
        field,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public async validateForm(form: IForm, method: string = "create", prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${form.model}`
    try {
      await validateForm(this.getFormData(form, prefix), form, {
        method,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public resetForm(form: IForm, prefix: string = "")  {
    iterateFields(form, (field) => {
      this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, undefined)
    })
  }
}

export default new FormStore()
