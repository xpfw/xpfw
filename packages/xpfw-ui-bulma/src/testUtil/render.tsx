import * as rtr from "preact-render-to-json"
import * as React from "react"

export default (Component: any, name: string) => {
  expect(rtr(Component)).toMatchSnapshot(name)
}
