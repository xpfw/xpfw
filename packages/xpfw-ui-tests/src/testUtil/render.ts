import * as rtr from "preact-render-to-json"
import * as React from "react"
const r: any = rtr
export default (Component: any, name: string) => {
  expect(r(Component)).toMatchSnapshot(name)
}
