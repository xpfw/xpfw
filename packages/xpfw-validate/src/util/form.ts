import { isObject } from "lodash"
import { IField, IForm } from "../typeDef"

const getFieldsFromForm: (form: IForm) => IField[] = (form: IForm) => {
  const fields: IField[] = []
  if (isObject(form)) {
    for (const section of form.sections) {
      for (const field of section.fields) {
        fields.push(field)
      }
    }
  }
  return fields
}

export { getFieldsFromForm }
