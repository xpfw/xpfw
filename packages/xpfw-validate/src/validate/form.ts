import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
