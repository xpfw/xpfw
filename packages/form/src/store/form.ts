import { get, set } from "lodash"
import { action, observable } from "mobx"
import { prependPrefix } from "../util/prefixMaker"

/**
 * Holds data relevant for Forms such as values and validation errors
 */
export class FormStoreClass {
  /**
   * MobX Observable containing all form data
   */
  @observable
  public formData: any = {}
  /**
   * MobX Observable containing all form errors
   */
  @observable
  public errors: any = {}
  /**
   * MobX Observable containing Loading states
   */
  @observable
  public loading: any = {}

  /**
   * Acquire a value from the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   * @param defaultValue value to return if it is not set in the form data
   */
  public getValue(valuePath?: string, prefix?: string, defaultValue?: any): any {
    return get(this.formData, prependPrefix(valuePath, prefix), defaultValue)
  }

  /**
   * Set a value in the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setValue(valuePath?: string, value?: any, prefix?: string) {
    set(this.formData, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getError(valuePath?: string, prefix?: string): any {
    return get(this.errors, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setError(valuePath?: string, value?: any, prefix?: string) {
    set(this.errors, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getLoading(valuePath?: string, prefix?: string): any {
    return get(this.loading, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setLoading(valuePath: string, value: boolean, prefix?: string) {
    set(this.loading, prependPrefix(valuePath, prefix), value)
  }
}

const FormStore = new FormStoreClass()
export default FormStore
