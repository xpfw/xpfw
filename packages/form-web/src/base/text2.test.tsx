import { ComponentRegistry, FormStore, SharedField } from "@xpfw/form"
import { DateField } from "@xpfw/form-tests"
import * as MockDate from "mockdate"
import * as React from "react"
import render from "../testUtil/render"
import TextField, { setDate } from  "./text"
MockDate.set(new Date(1304605243337))

ComponentRegistry.registerComponent("string", TextField)
ComponentRegistry.registerComponent("number", TextField)

test("Text Field Test", () => {
  const n: any = null
  let setTo = null
  render(<SharedField schema={DateField} />, "empty datetime-local field")
  const dateSetter = setDate((a: any) => {
    setTo = a
    FormStore.setValue(String(DateField.title), a)
  }, DateField, "value")
  expect(setTo).toBeNull()
  DateField.format = "date-time"
  dateSetter({value: "2018-02-22T07:55"})
  expect(setTo).toMatchSnapshot("after datetime-local set")
  render(<SharedField schema={DateField} />, "datetime-local with content")
  DateField.format = "date"
  dateSetter({value: "2012-01-11"})
  expect(setTo).toMatchSnapshot("after date set")
  render(<SharedField schema={DateField} />, "date with content")
  DateField.format = "time"
  dateSetter({value: "16:20"})
  expect(setTo).toMatchSnapshot("after time set")
  render(<SharedField schema={DateField} />, "time with content")
})
