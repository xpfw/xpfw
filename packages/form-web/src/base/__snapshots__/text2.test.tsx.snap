// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Field Test: after date set 1`] = `"2012-01-10T23:00:00.000Z"`;

exports[`Text Field Test: after datetime-local set 1`] = `"2018-02-21T23:00:00.000Z"`;

exports[`Text Field Test: after time set 1`] = `"2011-05-05T14:20:00.000Z"`;

exports[`Text Field Test: date with content 1`] = `
<input
  onChange={[Function]}
  type="date"
  value="2012-01-11"
/>
`;

exports[`Text Field Test: datetime-local with content 1`] = `
<input
  onChange={[Function]}
  type="datetime-local"
  value="2018-02-22T00:00"
/>
`;

exports[`Text Field Test: empty datetime-local field 1`] = `
<input
  onChange={[Function]}
  type="date"
  value="2011-05-05"
/>
`;

exports[`Text Field Test: time with content 1`] = `
<input
  onChange={[Function]}
  type="time"
  value="16:20"
/>
`;
